### ISSUE TITLE
Fix DDP synchronization: Add window broadcasting and sequential sampling mode

### ISSUE THREAD

**Issue by azadeh-gh** [NOAA-EMC/ocelot#73]
# Callback Comparison: Main Branch vs Current Implementation

### 1. DDP Synchronization

| Feature | Main Branch | Current Implementation |
|---------|---------------|---------------------------|
| **Window broadcast** | Each rank picks independently | Rank-0 picks, broadcasts to all |
| **Consistency** | Different ranks use different windows |  All ranks use same window |
| **Helper functions** | None |  `_broadcast_window()`, `_is_rank0()`, etc. |

#### Main Branch  
```python
def on_train_epoch_end(self, trainer, pl_module):
    """Prepare NEXT epoch's TRAIN window."""
    total_days = (self.train_end_date - self.train_start_date).days
    if self.mode == "sequential":
        # ... sequential logic ...
        start = self._seq_cursor
    else:
        # PROBLEM: Each rank generates its own random offset!
        offset = self._rand_offset(total_days, self.train_window.days)
        start = self.train_start_date + pd.Timedelta(days=offset)

    # Each rank sets different window!
    self._update_datamodule(trainer, start, end, is_train=True)
```

**Problem**: In random mode, each rank calls `random.randint()` independently, getting different windows!

```
Epoch 1:
- Rank 0: Uses window 2024-03-15 to 2024-03-27
- Rank 1: Uses window 2024-05-10 to 2024-05-22  ← DIFFERENT!
- Rank 2: Uses window 2024-07-03 to 2024-07-15  ← DIFFERENT!
```
# Why Synchronized Windows Are Critical for DDP Training

**DDP assumes all ranks train on the SAME dataset, just different samples.** When ranks use different time periods, they're effectively training on **different datasets**, which violates DDP's fundamental assumption and causes:
1. **Inconsistent gradients** (averaging gradients from different data distributions)
2. **Biased model updates** (some time periods over-represented)
3. **Training instability** (conflicting gradient signals)

---

### Correct DDP Behavior 

DDP training follows this pattern:

```
Step 1: Each rank processes DIFFERENT SAMPLES from SAME DATASET
┌─────────────────────────────────────────────────────────┐
│ Dataset: All bins from 2024-03-15 to 2024-03-27         │
│ Bins: bin1, bin2, bin3, ..., bin23                      │
└─────────────────────────────────────────────────────────┘
         ↓                ↓                ↓
    Rank 0           Rank 1           Rank 2
    ------           ------           ------
  bin1, bin2       bin3, bin4       bin5, bin6
  bin7, bin8       bin9, bin10      bin11, bin12

Step 2: Each rank computes gradients on its samples
    ∇₀ = gradients from Rank 0's bins
    ∇₁ = gradients from Rank 1's bins
    ∇₂ = gradients from Rank 2's bins

Step 3: AllReduce averages gradients across all ranks
    ∇_avg = (∇₀ + ∇₁ + ∇₂) / 3

Step 4: All ranks apply the SAME averaged gradient
    θ_new = θ_old - lr × ∇_avg
```

**Key principle**: Different samples, **same underlying data distribution**.

### Incorrect Behavior with Unsynchronized Windows 

```
Step 1: Each rank processes samples from DIFFERENT DATASETS
┌──────────────────────┐  ┌──────────────────────┐  ┌──────────────────────┐
│ Rank 0 Dataset:      │  │ Rank 1 Dataset:      │  │ Rank 2 Dataset:      │
│ 2024-03-15 to 03-27  │  │ 2024-06-10 to 06-22  │  │ 2024-09-05 to 09-17  │
│ (Spring data)        │  │ (Summer data)        │  │ (Fall data)          │
└──────────────────────┘  └──────────────────────┘  └──────────────────────┘
         ↓                          ↓                          ↓
    bin1, bin2                 bin1, bin2                 bin1, bin2
  (Spring bins)              (Summer bins)               (Fall bins)

Step 2: Each rank computes gradients on DIFFERENT DISTRIBUTIONS
    ∇₀ = gradients from Spring weather patterns
    ∇₁ = gradients from Summer weather patterns
    ∇₂ = gradients from Fall weather patterns

Step 3: AllReduce averages INCOMPATIBLE gradients
    ∇_avg = (∇_spring + ∇_summer + ∇_fall) / 3  ← MEANINGLESS!

Step 4: Model receives contradictory update signal
    θ_new = θ_old - lr × ∇_confused
```

**The problem**: Averaging gradients from **different data distributions** creates a meaningless update direction.

---

### 1. Violates DDP's Core Assumption

**DDP Assumption**:
> "All ranks are training on random samples from the same dataset."

**What this means**:
- Data distribution is the same across ranks
- Only the specific samples differ
- Averaging gradients is statistically valid

**When violated**:
- Each rank sees a different data distribution
- Gradients represent different optimization objectives
- Averaging them is like averaging apples and oranges

### 2. Inconsistent Gradient Signals

Imagine training a weather prediction model:

**Rank 0 (Spring data - March)**:
- Learns: "Temperature usually increases during the day"
- Learns: "Precipitation is moderate"
- Gradient pushes: ∇_temp ≈ +0.5, ∇_precip ≈ +0.2

**Rank 1 (Summer data - June)**:
- Learns: "Temperature is high and stable"
- Learns: "Precipitation is low"
- Gradient pushes: ∇_temp ≈ -0.3, ∇_precip ≈ -0.4

**Rank 2 (Fall data - September)**:
- Learns: "Temperature is decreasing"
- Learns: "Precipitation is variable"
- Gradient pushes: ∇_temp ≈ -0.2, ∇_precip ≈ +0.1

**After AllReduce**:
```
∇_avg_temp = (+0.5 - 0.3 - 0.2) / 3 = 0.0    ← Canceled out!
∇_avg_precip = (+0.2 - 0.4 + 0.1) / 3 = -0.03 ← Weak mixed signal
```

**Result**: Model receives weak, contradictory signals. It learns nothing effectively.

### 3. Biased Sampling

**Correct DDP**: Each time period has equal probability of being selected across all GPUs.

**Incorrect (unsynchronized)**: Some periods appear more often than others.

**Example** (8 ranks, 1 year of data):

**Synchronized (Correct)**:
```
All ranks use: March 15-27
- March 15-27: 8 samples (all ranks)
- Coverage: Uniform across selected window
```

**Unsynchronized (Broken)**:
```
Random selections by each rank:
- Rank 0: March 15-27
- Rank 1: March 20-Apr 1    ← Overlaps!
- Rank 2: June 10-22
- Rank 3: March 18-30       ← Overlaps!
- ...

Result:
- March data: Over-sampled (appears on ranks 0, 1, 3)
- April-May data: Under-sampled
- Some periods: Not sampled at all this epoch
```

**Impact**: Model becomes biased toward over-represented time periods.

### 4. Loss Becomes Meaningless

**Correct DDP Loss**:
```
Epoch 1, Window: 2024-03-15 to 03-27 (same for all ranks)

Rank 0: bin1, bin2, bin3     → Loss₀ = 0.42
Rank 1: bin4, bin5, bin6     → Loss₁ = 0.38
Rank 2: bin7, bin8, bin9     → Loss₂ = 0.45

Average Loss = (0.42 + 0.38 + 0.45) / 3 = 0.417

This number is meaningful: Average prediction error for March 15-27
```

**Incorrect (Unsynchronized)**:
```
Epoch 1

Rank 0: March 15-27          → Loss₀ = 0.42  (Spring weather)
Rank 1: June 10-22           → Loss₁ = 0.38  (Summer weather)
Rank 2: September 5-17       → Loss₂ = 0.45  (Fall weather)

Average Loss = (0.42 + 0.38 + 0.45) / 3 = 0.417

This number is MEANINGLESS: What does "average error across Spring, Summer, Fall" mean?
```

**Why meaningless?**:
- Can't track if model is improving on specific patterns
- Can't compare across epochs (different time period mixes)
- Can't debug (which period is causing high loss?)
- Loss could go down just because easier periods were randomly selected

---
#### Current Implementation (CORRECT for DDP)
```python
def on_train_epoch_end(self, trainer, pl_module):
    dm = trainer.datamodule
    used_start = pd.to_datetime(dm.hparams.train_start)

    # Only rank-0 picks the window
    if _is_rank0():
        if self.mode == "sequential":
            start = used_start + self.seq_stride
            # ...
        else:
            # Only rank-0 generates random offset
            offset = self._rand_offset(total_days, self.train_window.days)
            start = self.train_start_date + pd.Timedelta(days=offset)
            end = self._clip_end(start + self.train_window, self.train_end_date)
    else:
        # Other ranks use dummy values
        start = self.train_start_date
        end = self._clip_end(start + self.train_window, self.train_end_date)

    # Broadcast from rank-0 to all ranks
    start, end = _broadcast_window(start, end)

    # All ranks now have the SAME window
    self._update_datamodule(trainer, start, end, is_train=True)
```

**Result**: All ranks use the same window!

```
Epoch 1:
- Rank 0: Uses window 2024-03-15 to 2024-03-27  ✓
- Rank 1: Uses window 2024-03-15 to 2024-03-27  ✓ SAME
- Rank 2: Uses window 2024-03-15 to 2024-03-27  ✓ SAME
```

---

### 2. Sequential Mode Window Advancement

| Feature | Main Branch | Current Implementation  |
|---------|---------------|---------------------------|
| **Basis for next window** | Cursor (may desync) | Actual used window from DM |
| **Restart safety** | Cursor can be stale | Uses actual DM state |

#### Main Branch
```python
def on_train_epoch_end(self, trainer, pl_module):
    if self.mode == "sequential":
        if self._seq_cursor is None:
            self._seq_cursor = self.train_start_date
        start = self._seq_cursor  # ← Uses internal cursor
        # ...
        # advance cursor for next epoch
        next_start = start + self.seq_stride
        self._seq_cursor = next_start  # ← Updates cursor
```

**Problem**: If training restarts or dataloaders reload, `_seq_cursor` might not match actual DM state.

#### Current Implementation
```python
def on_train_epoch_end(self, trainer, pl_module):
    dm = trainer.datamodule
    used_start = pd.to_datetime(dm.hparams.train_start)  # ← Uses actual DM state

    if _is_rank0():
        if self.mode == "sequential":
            start = used_start + self.seq_stride  # ← Based on what was actually used
```

**Benefit**: Always advances from the **actual** window that was just used, preventing desync.

---

### 3. Logging and Debugging

| Feature | Main Branch | Current Implementation   |
|---------|-------------|---------------------------|
| **Rank identification** | `os.environ.get("RANK")` | Proper DDP `dist.get_rank()` |
| **Detailed headers** | Basic | `log_rank_header()` with GPU/node info |
| **Sync confirmation** | No | Shows "(sync)" after broadcast |

#### Main Branch
```python
def on_train_epoch_start(self, trainer, pl_module):
    rank = int(os.environ.get("RANK", "0"))  # ← Fragile
    print(f"[Rank {rank}] train range: ...")
```

#### Current Implementation
```python
def on_train_epoch_start(self, trainer, pl_module):
    if trainer.is_global_zero:
        log_rank_header("TRAIN", pl_module.current_epoch)
        # Shows: [GRank 0 | LRank 0 | Node 0 | GPU 0] === TRAIN EPOCH 5 START ===
```

**Better diagnostics** for multi-GPU/multi-node debugging.

---

### 4. Checkpoint Persistence

| Feature | Main Branch | Current Implementation  |
|---------|---------------|---------------------------|
| **Sequential cursor** | Has it | Has it |
| **Used window tracking** | No | Tracks `_last_used_start` |

Both implement `state_dict()` and `load_state_dict()` for checkpoint resuming, but current version tracks more state for better restart behavior.

---

### 5. SequentialDataCallback

Both files have a `SequentialDataCallback` class, but they differ:

#### Main Branch
```python
class SequentialDataCallback(pl.Callback):
    def __init__(self, full_start_date, full_end_date, window_days=7, stride_days=1):
        # No DDP synchronization
        # No mode parameter (always sequential)
        # No wrap control
```

#### Current Implementation
```python
class SequentialDataCallback(pl.Callback):
    def __init__(
        self,
        full_start_date,
        full_end_date,
        window_days=7,
        stride_days=1,
        mode="sequential",        # Can be "random" or "sequential"
        wrap_sequential=True,     # Control wraparound
    ):
        # Has DDP broadcast in on_train_epoch_end()
        # Tracks used window like ResampleDataCallback
```

**Current version is more flexible and DDP-safe.**
